# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()



word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

I predict the Random Forests Classifier model to perform better than the Logistical Regression Model for the Spam data set.  This is because the RFC model uses trees to predict outcomes, whereas LR model uses linear clssification for its prediction.  Spam is more amenable to trees level classfiication than linear classification.  Hence, I predict RFC model to be better.

## Split the Data into Training and Testing Sets

In [3]:
# Check if 'spam' column exists in the DataFrame
if 'spam' not in data.columns:
    print("Column 'spam' does not exist in the DataFrame.")
else:
    # Create the labels set `y` and features DataFrame `X`
    y = data['spam']
    X = data.drop(columns=['spam'])
    



In [4]:
# # Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [5]:

# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
X_train.head()
    







word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
958             0.00               0.00           0.00           0.0   
1533            0.00               0.00           0.00           0.0   
654             0.33               0.00           0.33           0.0   
1497            0.52               1.31           0.26           0.0   
3844            0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
958             0.0            0.00              1.21                0.00   
1533            0.0            0.00              0.00                0.00   
654             0.0            0.00              0.66                0.00   
1497            2.9            0.26              0.79                0.26   
3844            0.0            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
958               0.0            1.21  ...                   0.0          0.0   
1533              0.0            0.00  ...                   0.0          0.0   
654               0.0            0.00  ...                   0.0          0.0   
1497              0.0            0.79  ...                   0.0          0.0   
3844              0.0            0.00  ...                   0.0          0.0   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
958         0.000          0.0        0.000        0.000        0.000   
1533        0.208          0.0        0.417        0.208        0.000   
654         0.116          0.0        0.406        0.464        0.348   
1497        0.138          0.0        0.046        0.000        0.000   
3844        0.000          0.0        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
958                        1.676                          15   
1533                       3.812                          16   
654                        6.932                          43   
1497                       2.934                          60   
3844                      14.000                          53   

      capital_run_length_total  
958                         57  
1533                        61  
654                        513  
1497                       135  
3844                        56  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()




In [7]:


# Fit the Standard Scaler with the training data
X_Scaler = scaler.fit(X_train)

# Scale the training data
X_train_scaled = X_Scaler.transform(X_train)
X_test_scaled = X_Scaler.transform(X_test)


    





## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [12]:

from sklearn.linear_model import LogisticRegression
lr_model = LogisticRegression(random_state=1)
lr_model.fit(X_train_scaled, y_train)
y_pred = lr_model.predict(X_test_scaled)
y_pred










array([0, 0, 0, ..., 1, 1, 0])

In [13]:
accuracy_score(y_test, y_pred)



0.9232440260680667

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [14]:

# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(random_state=1)
rf_model.fit(X_train_scaled, y_train)
y_pred_rf = rf_model.predict(X_test_scaled)
y_pred_rf


array([0, 0, 0, ..., 1, 1, 0])

In [16]:
accuracy_score(y_test, y_pred_rf)
    
    


0.9572773352643013

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*Replace the text in this markdown cell with your answers to these questions.*

The accuracy of the Random Forest Classifier model is 0.9572773352643013

The accuracy of the Logistic Regression model is 0.9232440260680667

The Random Forest Classifier model has a higher accuracy than the Logistic Regression model.

The Random Forest Classifier model is the best model for this dataset.

The performance of the RFC model compares favorably with the prediction that spam dataset analysis is likely to be more accurate with a decision trees level analysis than a linear data level analysis.  


    